lib/Db/TimelineQueryFaceRecognition.php

```diff
@@ -0,0 +1,222 @@
+        }
+
+        // ORDER by number of faces in cluster
+        $query->orderBy('name', 'ASC');
+        $query->addOrderBy('count', 'DESC');
+        $query->addOrderBy('frp.id'); // tie-breaker
+
+        // FETCH all faces
+        $cursor = $this->executeQueryWithCTEs($query);
+        $faces = $cursor->fetchAll();
+
+        // Post process
+        foreach ($faces as &$row) {
+            $row['id'] = $row['name'] ? $row['name'] : (int) $row['id'];
+            $row['count'] = (int) $row['count'];
+        }
+
+        return $faces;
+    }
+
+    public function getFaceRecognitionPreview(Folder &$folder, $previewId)
+    {
+        $query = $this->connection->getQueryBuilder();
+
+        // SELECT face detections
+        $query->select(
+            'fri.file as file_id',      // Get actual file
+            'frf.left',                 // Image cropping
+            'frf.top',
+            'frf.right',
+            'frf.bottom',
+            'm.w as image_width',       // Scoring
+            'm.h as image_height',
+            'frf.confidence',
+            'm.fileid',
+            'm.datetaken',              // Just in case, for postgres
+        )->from('facerecog_faces', 'frf');
+
+        // WHERE faces are from images.
+        $query->innerJoin('frf', 'facerecog_images', 'fri', $query->expr()->eq('fri.id', 'frf.image'));
+
+        // WHERE these photos are memories indexed
+        $query->innerJoin('fri', 'memories', 'm', $query->expr()->eq('m.fileid', 'fri.file'));
+
+        if (is_numeric($previewId)) {
+            // WHERE faces are from id persons (a cluster).
+            $query->innerJoin('frf', 'facerecog_persons', 'frp', $query->expr()->eq('frp.id', 'frf.person'));
+            $query->where($query->expr()->eq('frp.id', $query->createNamedParameter($previewId)));
+        } else {
+            // WHERE faces are from name on persons.
+            $query->innerJoin('frf', 'facerecog_persons', 'frp', $query->expr()->eq('frp.id', 'frf.person'));
+            $query->where($query->expr()->eq('frp.name', $query->createNamedParameter($previewId)));
+        }
+
+        // WHERE these photos are in the user's requested folder recursively
+        $query = $this->joinFilecache($query, $folder, true, false);
+
+        // LIMIT results
+        $query->setMaxResults(15);
+
+        // Sort by date taken so we get recent photos
+        $query->orderBy('m.datetaken', 'DESC');
+        $query->addOrderBy('m.fileid', 'DESC'); // tie-breaker
+
+        // FETCH face detections
+        $cursor = $this->executeQueryWithCTEs($query);
+        $previews = $cursor->fetchAll();
+        if (empty($previews)) {
+            return null;
+        }
+
+        // Score the face detections
+        foreach ($previews as &$p) {
+            // Get actual pixel size of face
+            $iw = min((int) ($p['image_width'] ?: 512), 2048);
```

Comment:
Any way this code could be de-duplicated? I assume the logic is unchanged/similar?

Reply:
Face Recognition saves the detections as top, bottom, right, left. I have intentions to change to width, height, but it will never be compatible with recognize, which uses these parameters as percentages of the image.

This is probably one of the biggest recognize errors, since the detections must be pixel perfect when getting the face alignments and descriptors.
However, I understand that this decision is due to the fact that it is what face-api offers.

So, I will always have to calculate the percentages, but (except the confidence, which is a score in itself) the subsequent calculation of the score could be reused.

Reply:
Converting them either way is fine -- if it makes more sense, you can convert recognize's percent values to absolute ones and call the common function.

Reply:
Partially done..
https://github.com/matiasdelellis/facerecognition/pull/608/commits/bfdaab2cd24c58b737c65351e02493fc33055d54

Now I have to adapt the queries.